### PR TITLE
[2/10] CMSIS-NN code generator for softmax

### PR DESCRIPTION
**Pull request by ashutosh-arm**
This PR is to support softmax (int8) via CMSIS-NN. It also includes changes for CMSIS-NN graph partitioner since that PR has not merged yet: https://github.com/apache/tvm/pull/8653

This includes second and third steps of plan mentioned in the RFC: https://github.com/apache/tvm-rfcs/pull/15

Co-authored-by: Chris Sidebottom chris.sidebottom@arm.com

### CODE REVIEW

**Comment by ashutosh-arm** [comment]
Adding @manupa-arm @grant-arm @areusch for review.

File: python/tvm/relay/backend/contrib/cmsisnn/codegen.py
Comment (Mousius) [inline python/tvm/relay/backend/contrib/cmsisnn/codegen.py:126]: `@tvm.register_func`

File: python/tvm/relay/backend/contrib/cmsisnn/codegen.py
Comment (Mousius) [inline python/tvm/relay/backend/contrib/cmsisnn/codegen.py:133]: What's the purpose of this assertion?

File: src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:82]: Is there a more concrete name we can provide here? Like `cmsisnn_softmax` ?

File: src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc
Comment (Mousius) [inline src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:32]: I'm not sure all of these headers are needed in this file? For example, `tvm/runtime/ndarray.h` is also included from `tvm/tir/function.h`. Can you limit these to specifically what's being used in this file?

File: tests/python/contrib/test_cmsisnn/test_softmax.py
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_softmax.py:34]: Can you not use `from tests.python.relay.aot.aot_test_utils import ...` ?

File: tests/python/contrib/test_cmsisnn/test_softmax.py
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_softmax.py:164]: Can we test `scale` and `zero_point` independently and in combination with data types? This test could fail to match based on one or many of these.

**Review by Mousius — CHANGES_REQUESTED** [review]
This is looking great @ashutosh-arm, really enjoying the code re-use here :smile_cat: Just needs neatening up a bit from my POV.

File: python/tvm/relay/backend/contrib/cmsisnn/codegen.py
Comment (ashutosh-arm) [reply python/tvm/relay/backend/contrib/cmsisnn/codegen.py:26]: ACK

File: tests/python/contrib/test_cmsisnn/test_softmax.py
Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_softmax.py:34]: ACK

File: python/tvm/relay/backend/contrib/cmsisnn/codegen.py
Comment (ashutosh-arm) [reply python/tvm/relay/backend/contrib/cmsisnn/codegen.py:126]: ACK. Removed additional _ffi_api.py associated with it.

File: python/tvm/relay/backend/contrib/cmsisnn/codegen.py
Comment (ashutosh-arm) [reply python/tvm/relay/backend/contrib/cmsisnn/codegen.py:133]: Result of copy-n-paste. But my guess is that non AOT flows might be supplying other types too.

File: src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:82]: The name here is arm_softmax_s8. I have updated the string with that.

File: src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:32]: Thanks for pointing that out. I had it in mind as well. Now, the list has been reduced to just 3 includes :smile:

File: tests/python/contrib/test_cmsisnn/test_softmax.py
Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_softmax.py:164]: Done!

File: python/tvm/relay/backend/contrib/cmsisnn/codegen.py
Comment (Mousius) [reply python/tvm/relay/backend/contrib/cmsisnn/codegen.py:133]: I'm not convinced anything other than a function will come here, but if you're concerned can we have a test case that this assertion is raised?

Comment (ashutosh-arm) [reply python/tvm/relay/backend/contrib/cmsisnn/codegen.py:133]: Ok, I will remove it then.

**Review by Mousius — CHANGES_REQUESTED** [review]
One small suggestion and I think you need to run the linters against the code again @ashutoshparkhi :smile_cat:

File: src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc
Comment (areusch) [inline src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:102]: @jroesch @electriclilies @tqchen what's our stance on adding this much complexity to the codegen? can you comment whether it's okay to invoke `tvm::relay::qnn::GetFixedPointMultiplierShift` from here?

in general, we should favor putting as much validation logic in the compiler as possible. it kind of seems like this should move upwards though, so that the codegen is just emitting tir.call_extern with the same arguments as are in the CallNode? Doing that would make it possible to emit code using llvm codegen.

File: src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc
Comment (areusch) [inline src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:111]: is there expected to be only one CMSIS-NN call per primfunc?

**Review by areusch — COMMENTED** [review]
@ashutosh-arm thanks for the PR. a few questions on the approach

File: src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:111]: At the moment, yes. One external Relay function to one Primfunc to one CMSIS-NN API is what we were thinking initially.

File: src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:102]: For the second point, we were planning on moving TIR generation from python implementation to C++ when we go to next operator's support. I can move this call to TIR then if others are okay with that.

Comment (areusch) [reply src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:102]: @ashutosh-arm @Mousius and I discussed about this a bit. This is a smaller instance of a more general problem: other CMSIS-NN operators require user-defined data structures. This is the _only_ thing stopping them from being encoded as `tir.call_extern` and therefore allowing reuse of the `c` and `llvm` codegen.

@ZihengJiang was working on this but I believe it's been de-prioritized. my preference is not to place call-specific transformation logic inside of codegen and instead ensure the `tir.CallNode` contains the final parameters where possible. might it be possible to work around this either:
a) by creating a custom TIR Node class and using this as a parameter, knowing that this parameter can only be consumed by CMSIS-NN codegen
b) by creating a placeholder parameter and attaching the struct as an attribute to the CallNode

cc @mbs-octoml

Comment (tqchen) [reply src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:102]: On the data structure bits. Note that most of the data structures are passed as pointers(handle), this means we likely only need intrinsics to construct those data structures(return handles), then pass as argument to the call.

We would prefer this way of reusing existing construct(via composing intrinsics) than adding semantics to the TIR when possible.

Comment (areusch) [reply src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:102]: @tqchen the issue with this approach is that we want to stack-allocate these structs because we don't want to call out to the memory manager. so while they're passed by pointer, we still need to account for the memory usage of those structs.

we probably need to discuss this a bit more before merging calls with structs, but I think we understand enough to unblock this PR. I discussed with @ashutosh-arm and he'll update this PR to move this call out of the codegen. The next CMSIS-NN call will require us to fill a struct. This way, when we move to consider that call, we'll start from a clean slate where the checked-in codegen doesn't further require a cleanup to meet the intended codegen pattern and therefore remove that from the discussion.

Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc:102]: Since this PR has already been reviewed. I've created another one here: #8951 that contains changes to move TIR code generation to C++ side. This PR also has the change to move out the CMSIS-NN Softmax's parameter calculations out of C codegen.

**Review by areusch — APPROVED** [review]

**Comment by areusch** [comment]
thanks @ashutosh-arm for working through this with us! we will continue review in the next PR.

**Pull request merged**
